hilti/toolchain/src/compiler/printer.cc

```diff
@@ -1111,20 +1118,26 @@ void printer::print(std::ostream& out, Node* root, bool compact) {
     }
     else {
         auto stream = printer::Stream(out);
-        print(stream, root);
+        stream._print(root);
     }
 }
 
-void printer::print(printer::Stream& stream, Node* root) {
+void printer::Stream::_print(Node* root) {
     util::timing::Collector _("hilti/printer");
 
+    struct _ {
+        ~_() { detail::State::current->current_plugin = nullptr; }
+    } __;
+
```

Comment:
We do almost the same in `printer::print` already but for the check for `detail::State::depth`, and if I remove this code here the test suite still passes, so not sure why we need it.

In any case I think not having this management code here in `_print` but in `print` would be much easier to follow since `current_plugin` gets set up there as well (otherwise calling `_print` would always reset `current_plugin` as a side-effect).

Reply:
It's technically probably not necessary, but seems cleaner not leave the state set to something that's not longer valid/relevant.

> In any case I think not having this management code here in _print but in print would be much easier to follow since current_plugin gets set up there as well 

`current_plugin` gets set only in `_print` actually, just before we execute the plugin's hook, so I'd leave the clearing there as well.

Btw, I think we should add a utility class sometime that's like this: https://github.com/zeek/zeek-agent-v2/blob/7e07589bc81d7ee6f49b703ffad13239c1ab7b8e/src/util/helpers.h#L36-L49

Reply:
I got confused :confused:; `print` sets up the current _state_ via `current` (so it can e.g., `setCompact` which is in `State`, and `print_` sets _`current_plugin`_ which lives in the `current` state. Sorry for the noise.

The way the cleanup works here we still leak `current_plugin` from the loop, maybe like this:

```cpp
void printer::Stream::_print(Node* root) {
    util::timing::Collector _("hilti/printer");

    for ( auto& p : plugin::registry().plugins() ) {
        if ( ! p.ast_print )
            continue;

        auto* prev = std::exchange(detail::State::current->current_plugin, &p);
        auto _ = scope_exit([&]() { detail::State::current->current_plugin = prev; });

        if ( (*p.ast_print)(root, *this) )
            return;
        else {
            // If the print hook did not succeed defer to default printer.
            // This might still make use of the currently selected plugin.
            Printer(*this).dispatch(root);
            return;
        }
    }

    // Defer to the default printer with the current plugin (which might be unset).
    //
    // TODO: We should be able to replace this with `util::cannotBeReached()`
    // if the HILTI plugin provides `ast_print` (e.g., which could just return `false`).
    Printer(*this).dispatch(root);
}
```

This uses a hypothetical scope guard a la [`std::experimental::scope_exit`](https://en.cppreference.com/w/cpp/experimental/scope_exit)

```cpp

template<typename EF>
struct scope_exit {
    scope_exit(EF f) noexcept : _f(std::forward<EF>(f)) {}

    scope_exit(const scope_exit&) = delete;
    scope_exit(scope_exit&&) = delete;

    ~scope_exit() { _f(); }

    EF _f;
};
```

Reply:
>  we still leak current_plugin from the loop

Oh, thanks for that observation! Turns out that demonstrates something unexpected: if I change it to not leak, the PR doesn't work anymore. The thing is that the `dispatch()` still needs to get that new ID hook to execute when we're printing Spicy code. But when executing, it checks the current plugin and if that's now unset, the hook won't do anything. Need to restructure that a bit once more.